### PR TITLE
[FLINK-12254][table-common] Preparation for using the new type system

### PR DESCRIPTION
**Pull request by twalthr**
## Brief change log

This PR is a preparation for FLINK-12254. It fixes a couple of shortcomings when working with the new type system and adds the first utility classes that might useful for other contributions as well.

## Verifying this change

See commit messages.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/VarCharType.java [hunk 1]
```diff
@@ -86,6 +89,9 @@ public LogicalType copy(boolean isNullable) {
 
 	@Override
 	public String asSerializableString() {
+		if (length == MAX_LENGTH) {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/VarCharType.java]: Could we have that only in the `summaryString`? Personally I'd prefer to serialize only the actual types and not the shortcuts.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/ClassDataTypeConverter.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	/**
+	 * Returns the clearly identifiable data type if possible. For example, {@link Long} can be
+	 * expressed as {@link DataTypes#BIGINT()}. However, for example, {@link Row} cannot be extracted
+	 * as information about the fields is missing. Or {@link BigDecimal} needs to be mapped from a
+	 * variable precision/scale to constant ones.
+	 */
+	public static Optional<DataType> extractDataType(Class<?> clazz) {
+		// byte arrays have higher priority that regular arrays
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/ClassDataTypeConverter.java]:
```suggestion
		// byte arrays have higher priority than regular arrays
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/DataTypeVisitor.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * The visitor definition of {@link DataType}. The visitor transforms a data type into
+ * instances of {@code R}.
+ */
+@PublicEvolving
+public interface DataTypeVisitor<R> {
+
+	R visit(AtomicDataType atomicDataType);
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/DataTypeVisitor.java]: Makes sense to me. Will update the visitor methods.

[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/DataTypeVisitor.java]: We also not do it for the LogicalTypeVisitor and it's also redundant. Let's keep it as it is.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/VarCharType.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/VarCharType.java]: I was also unsure about this change. Summary strings should be sufficient, I agree.

**Pull request closed without merging**